### PR TITLE
[C++ API] Distance module

### PR DESCRIPTION
**Pull request by jon-tow**
Adds `Distance` module parity. 
#25883

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/options/distance.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+#include <torch/arg.h>
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/types.h>
+
+namespace torch {
+namespace nn {
+
+/// Options for the `CosineSimilarity` module.
+struct TORCH_API CosineSimilarityOptions {
+  // Dimension where cosine similarity is computed. Default: 1
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/distance.h]: nit: `///` instead of `//`

File: torch/csrc/api/include/torch/nn/options/distance.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  TORCH_ARG(int64_t, dim) = 1;
+  // Small value to avoid division by zero. Default: 1e-8
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/distance.h]: nit: `///` instead of `//`

File: torch/csrc/api/include/torch/nn/modules/distance.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#include <torch/types.h>
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+
+namespace torch {
+namespace nn {
+
+/// Returns the cosine similarity between :math:`x_1` and :math:`x_2`, computed
+/// along `dim`.
+struct TORCH_API CosineSimilarityImpl : Module {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/distance.h]: we should subclass `CosineSimilarityImpl` from `public torch::nn::Cloneable<CosineSimilarityImpl>`, to match the other C++ modules.

File: torch/csrc/api/include/torch/nn/modules/distance.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+/// See the documentation for `CosineSimilarityImpl` class to learn what methods
+/// it provides, or the documentation for `ModuleHolder` to learn about
+/// Pytorch's module storage semantics.
+TORCH_MODULE(CosineSimilarity);
+
+// ============================================================================
+
+/// Returns the batchwise pairwise distance between vectors :math:`v_1`,
+/// :math:`v_2` using the p-norm.
+struct TORCH_API PairwiseDistanceImpl : Module {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/distance.h]: we should subclass `PairwiseDistanceImpl` from `public torch::nn::Cloneable<PairwiseDistanceImpl>`, to match the other C++ modules.

File: torch/csrc/api/include/torch/nn/functional/distance.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+namespace functional {
+
+inline Tensor cosine_similarity(
+    const Tensor& x1,
+    const Tensor& x2,
+    const CosineSimilarityOptions& options) {
+  return torch::cosine_similarity(
+      x1,
+      x2,
+      options.dim_,
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/distance.h]: We should use `options.dim()` instead of `options.dim_`, because `options.dim_` will be made private in the near future (https://github.com/pytorch/pytorch/pull/26419) (similarly for other options args as well).

File: test/cpp_api_parity/torch_nn_modules.py [hunk 6]
```diff
@@ -148,8 +148,14 @@
...
-    'CosineSimilarity': TorchNNModuleMetadata(),
-    'PairwiseDistance': TorchNNModuleMetadata(),
+    'CosineSimilarity': TorchNNModuleMetadata(
+        cpp_default_constructor_args="()",
+        num_attrs_recursive=2,
+    ),
+    'PairwiseDistance': TorchNNModuleMetadata(
+        cpp_default_constructor_args="()",
+        num_attrs_recursive=3,
+    ),
```
Comment (yf225) [inline test/cpp_api_parity/torch_nn_modules.py]: Thanks for updating the entries here! The parity test mechanism is still a bit fragile and I think it's okay to leave the entries unchanged here for now (mostly because I don't want contributors to have to worry about parity test ;)). I will add the parity test in a separate PR after we merge this PR.

File: torch/csrc/api/src/nn/modules/distance.cpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+namespace F = torch::nn::functional;
+
+namespace torch {
+namespace nn {
+
+CosineSimilarityImpl::CosineSimilarityImpl(CosineSimilarityOptions options)
+    : options(options) {}
+
+void CosineSimilarityImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::CosineSimilarity()";
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/distance.cpp]: We probably want to print all options args here as well. An example is `pretty_print` in torch/csrc/api/src/nn/modules/conv.cpp.

File: torch/csrc/api/src/nn/modules/distance.cpp [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  return F::cosine_similarity(x1, x2, options);
+}
+
+// ============================================================================
+
+PairwiseDistanceImpl::PairwiseDistanceImpl(PairwiseDistanceOptions options)
+    : options(options) {}
+
+void PairwiseDistanceImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::PairwiseDistance()";
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/distance.cpp]: ditto - we probably want to print all options args here as well

File: test/cpp/api/modules.cpp [hunk 9]
```diff
@@ -532,6 +558,14 @@ TEST_F(ModulesTest, PrettyPrintEmbedding) {
       "torch::nn::Embedding(count=10, dimension=2)");
 }
 
+TEST_F(ModulesTest, PrettyPrintCosineSimilarity) {
+  ASSERT_EQ(c10::str(CosineSimilarity()), "torch::nn::CosineSimilarity()");
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: We should change the expected string after we change `pretty_print` of the new modules.

File: test/cpp/api/modules.cpp [hunk 10]
```diff
@@ -449,6 +449,32 @@ TEST_F(ModulesTest, L1Loss) {
...
+  CosineSimilarity cos(CosineSimilarityOptions().dim(1));
+  auto input1 = torch::randn({5, 3}, torch::requires_grad());
+  auto input2 = torch::randn({5, 3}, torch::requires_grad());
+  auto output = cos->forward(input1, input2);
+  auto s = output.sum();
+  s.backward();
+
+  ASSERT_EQ(output.ndimension(), 1);
+  ASSERT_EQ(output.size(0), 5);
+  ASSERT_EQ(input1.sizes(), input1.grad().sizes());
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: Is there a way to check the value of `output` (e.g. by assigning special values to `input1` and `input2`)?

File: test/cpp/api/modules.cpp [hunk 11]
```diff
@@ -449,6 +449,32 @@ TEST_F(ModulesTest, L1Loss) {
...
+  PairwiseDistance dist(PairwiseDistanceOptions(1));
+  auto input1 = torch::randn({6, 4}, torch::requires_grad());
+  auto input2 = torch::randn({6, 4}, torch::requires_grad());
+  auto output = dist->forward(input1, input2);
+  auto s = output.sum();
+  s.backward();
+
+  ASSERT_EQ(output.ndimension(), 1);
+  ASSERT_EQ(output.size(0), 6);
+  ASSERT_EQ(input1.sizes(), input1.grad().sizes());
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: ditto here: is there a way to check the value of `output` (e.g. by assigning special values to `input1` and `input2`)?

**Review by yf225 — COMMENTED** [review]
@jon-tow Thanks a lot for the PR and really appreciated it. I left some minor comments on the implementation.

**Review by yf225 — APPROVED** [review]
@jon-tow Thanks so much for the contribution and really appreciated the high-quality work. I will merge it today.

**Comment by jon-tow** [comment]
Glad to help out @yf225! Thank you for guiding me through it :).
